apps/api/pages/api/destination-calendars/[id]/_patch.ts

```diff
@@ -57,15 +60,130 @@ import { schemaQueryIdParseInt } from "~/lib/validations/shared/queryIdTransform
+  const { userId, isAdmin, prisma, query, body } = req;
   const { id } = schemaQueryIdParseInt.parse(query);
   const parsedBody = schemaDestinationCalendarEditBodyParams.parse(body);
+  let credentialId: number | undefined = undefined;
+  const assignedUserId = isAdmin ? parsedBody.userId || userId : userId;
+
+  // when linked with eventTypeId, we need to fetch the userId from the eventTypeId
+  if (parsedBody.eventTypeId) {
+    const eventType = await prisma.eventType.findFirst({
+      where: { id: parsedBody.eventTypeId, userId: assignedUserId },
+    });
+    if (!eventType)
+      throw new HttpError({
+        statusCode: 400,
+        message: "Bad request, eventTypeId invalid",
+      });
+    parsedBody.userId = undefined;
+  }
+
+  const { userCredentials } = await findUserCredentials(req);
+
+  if (parsedBody.integration && !parsedBody.externalId) {
+    throw new HttpError({ statusCode: 400, message: "External Id is required with integration value" });
+  }
+  if (!parsedBody.integration && parsedBody.externalId) {
+    throw new HttpError({ statusCode: 400, message: "Integration value is required with external ID" });
+  }
+
+  if (parsedBody.integration && parsedBody.externalId) {
+    const calendarCredentials = getCalendarCredentials(userCredentials);
+
+    const { connectedCalendars } = await getConnectedCalendars(
+      calendarCredentials,
+      [],
+      parsedBody.externalId
+    );
+    const connectedCalendar = connectedCalendars.find(
+      (c) =>
+        c?.primary?.externalId === parsedBody.externalId && c?.primary?.integration === parsedBody.integration
+    );
+
+    if (!connectedCalendar)
+      throw new HttpError({
+        statusCode: 400,
+        message: "Bad request, credential id invalid",
+      });
+
+    credentialId = connectedCalendar.primary?.credentialId;
+  }
 
   const destinationCalendar = await prisma.destinationCalendar.update({
     where: { id },
-    data: parsedBody,
+    data: { ...parsedBody, credentialId },
   });
   return { destinationCalendar: schemaDestinationCalendarReadPublic.parse(destinationCalendar) };
 }
 
+async function findUserCredentials(req: NextApiRequest) {
+  const { userId: requestUserId, prisma, isAdmin } = req;
+  const { id } = schemaQueryIdParseInt.parse(req.query);
+  const body = schemaDestinationCalendarEditBodyParams.parse(req.body);
+  const userId = isAdmin ? body.userId || requestUserId : requestUserId;
+  const destinationCalendarObject = await prisma.destinationCalendar.findFirst({
+    where: {
+      id,
+    },
+    select: { eventTypeId: true, credentialId: true },
+  });
+
+  if (!destinationCalendarObject) {
+    throw new HttpError({
+      statusCode: 404,
+      message: `Destination calendar with ID ${id} not found`,
+    });
+  }
+
+  if (!destinationCalendarObject.credentialId) {
+    throw new HttpError({
+      statusCode: 404,
+      message: `Destination calendar missing credential id`,
+    });
+  }
```

Comment:
If the destinationCalendar entry is not linked to a credential, we throw an error